File: apps/backend/app/models/pre_incapacidad.py
"""
Modelo SQLAlchemy para PreIncapacidad.
Registro plano de radicaciones recibidas desde el portal externo,
pendientes de procesamiento por job/tarea programada.
"""
from datetime import date
from decimal import Decimal
from typing import Optional, List
from sqlalchemy import String, Integer, Date, Numeric, Text, BigInteger, Sequence
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import BaseModel


class PreIncapacidad(BaseModel):
    """
    Registro de pre-radicación de incapacidad ARL.
    No crea una Incapacidad real — es procesada posteriormente por un job.
    """

    __tablename__ = "pre_incapacidad"

    # Número secuencial legible, inicia en 202600001
    numero_radicacion: Mapped[int] = mapped_column(
        Integer,
        Sequence("pre_incapacidad_numero_seq", start=202600001, increment=1),
        unique=True,
        nullable=False,
        index=True,
    )

    # Estado del registro
    estado: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default="PENDIENTE",
        index=True,
        comment="PENDIENTE | PROCESADA | RECHAZADA | ERROR | DEVUELTA",
    )

    # ── Solicitante (datos planos) ───────────────────────────────────────────
    solicitante_correo: Mapped[str] = mapped_column(String(255), nullable=False)
    solicitante_nombres: Mapped[str] = mapped_column(String(200), nullable=False)
    solicitante_apellidos: Mapped[Optional[str]] = mapped_column(String(200))
    solicitante_telefono: Mapped[Optional[str]] = mapped_column(String(20))

    # ── Empresa (datos planos — None si independiente) ────────────────────────
    empresa_nit: Mapped[Optional[str]] = mapped_column(String(20))
    empresa_nombre: Mapped[Optional[str]] = mapped_column(String(255))

    # ── Empleado (datos planos) ───────────────────────────────────────────────
    empleado_tipo_documento: Mapped[str] = mapped_column(String(10), nullable=False)
    empleado_numero_documento: Mapped[str] = mapped_column(String(20), nullable=False)
    empleado_nombres: Mapped[str] = mapped_column(String(200), nullable=False)
    empleado_apellidos: Mapped[Optional[str]] = mapped_column(String(200))
    empleado_email: Mapped[Optional[str]] = mapped_column(String(255))
    empleado_telefono: Mapped[Optional[str]] = mapped_column(String(20))

    # ── Datos de la incapacidad ───────────────────────────────────────────────
    tipo: Mapped[str] = mapped_column(String(10), nullable=False, default="ARL")
    tipo_enfermedad: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        comment="ACCIDENTE_TRABAJO | ENFERMEDAD_LABORAL | ACCIDENTE_TRAYECTO",
    )
    fecha_inicio: Mapped[date] = mapped_column(Date, nullable=False)
    fecha_fin: Mapped[date] = mapped_column(Date, nullable=False)
    dias_totales: Mapped[int] = mapped_column(Integer, nullable=False)
    diagnostico_cie10: Mapped[str] = mapped_column(String(10), nullable=False)
    descripcion_diagnostico: Mapped[Optional[str]] = mapped_column(Text)
    nombre_medico: Mapped[str] = mapped_column(String(200), nullable=False)
    registro_medico: Mapped[str] = mapped_column(String(50), nullable=False)
    ips: Mapped[Optional[str]] = mapped_column(String(255))
    valor_dia: Mapped[Optional[Decimal]] = mapped_column(Numeric(14, 2))
    observaciones: Mapped[Optional[str]] = mapped_column(Text)

    # ── Trazabilidad de procesamiento ─────────────────────────────────────────
    error_procesamiento: Mapped[Optional[str]] = mapped_column(Text)

    # ── Devolución (set by internal users) ───────────────────────────────────
    motivo_devolucion: Mapped[Optional[str]] = mapped_column(Text)

    # ── Relaciones ────────────────────────────────────────────────────────────
    documentos: Mapped[List["PreDocumento"]] = relationship(
        "PreDocumento",
        back_populates="pre_incapacidad",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    validation_inconsistencias: Mapped[List["ValidationInconsistencia"]] = relationship(
        "ValidationInconsistencia",
        back_populates="pre_incapacidad",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    def __repr__(self) -> str:
        return f"<PreIncapacidad {self.numero_radicacion} [{self.estado}]>"

File: apps/frontend/sistema-interno/src/types/preIncapacidad.ts
import type { ValidationInconsistenciaRead } from './validationInconsistencia';

export type EstadoPreIncapacidad =
  | 'PENDIENTE'
  | 'PROCESADA'
  | 'RECHAZADA'
  | 'DEVUELTA'
  | 'ERROR';

export interface PreIncapacidadListItem {
  id: string;
  numero_radicacion: number;
  estado: EstadoPreIncapacidad;
  tipo: 'ARL' | 'SALUD';
  empleado_nombres: string;
  empleado_numero_documento: string;
  empresa_nit: string | null;
  empresa_nombre: string | null;
  fecha_inicio: string;
  fecha_fin: string;
  dias_totales: number;
  total_errores: number;
  total_warnings: number;
  created_at: string;
}

export interface PreDocumento {
  id: string;
  tipo_documento: string;
  nombre_original: string;
  estado_subida: 'OK' | 'ERROR';
  created_at: string;
}

export interface PreIncapacidadDetalle {
  id: string;
  numero_radicacion: number;
  estado: EstadoPreIncapacidad;
  tipo: string;
  tipo_enfermedad: string;
  solicitante_correo: string;
  solicitante_nombres: string;
  solicitante_apellidos: string | null;
  solicitante_telefono: string | null;
  empresa_nit: string | null;
  empresa_nombre: string | null;
  empleado_tipo_documento: string;
  empleado_numero_documento: string;
  empleado_nombres: string;
  empleado_apellidos: string | null;
  empleado_email: string | null;
  fecha_inicio: string;
  fecha_fin: string;
  dias_totales: number;
  diagnostico_cie10: string;
  descripcion_diagnostico: string | null;
  nombre_medico: string;
  registro_medico: string;
  ips: string | null;
  valor_dia: string | null;
  error_procesamiento: string | null;
  motivo_devolucion: string | null;
  created_at: string;
  documentos: PreDocumento[];
  validation_inconsistencias: ValidationInconsistenciaRead[];
}

export interface PreIncapacidadUpdate {
  empresa_nit?: string;
  empresa_nombre?: string;
  empleado_tipo_documento?: string;
  empleado_numero_documento?: string;
  empleado_nombres?: string;
  empleado_apellidos?: string;
  empleado_email?: string;
  tipo_enfermedad?: string;
  fecha_inicio?: string;
  fecha_fin?: string;
  diagnostico_cie10?: string;
  descripcion_diagnostico?: string;
  nombre_medico?: string;
  registro_medico?: string;
  ips?: string;
  valor_dia?: number;
}

export interface DevolucionRequest {
  motivo: string;
}

export interface DevolucionResponse {
  id: string;
  estado: string;
  motivo_devolucion: string;
  email_enviado: boolean;
}

export interface PromocionResponse {
  success: boolean;
  pre_incapacidad_id: string;
  incapacidad_id: string | null;
  errors: number;
  warnings: number;
  message: string;
}

export interface BandejaFiltros {
  estado?: EstadoPreIncapacidad;
  search?: string;
  skip?: number;
  limit?: number;
}
